Fix panic on fetch provider initialization

blakerouse opened the pull request:
<!-- Type of change
Please label this PR with one of the following labels, depending on the scope of your change:
- Bug
- Enhancement
- Breaking change
- Deprecation
- Cleanup
- Docs
-->

## What does this PR do?

<!-- Mandatory
Explain here the changes you made on the PR. Please explain the WHAT: patterns used, algorithms implemented, design architecture, message processing, etc.
-->

Fixes an issue where the initialization of a fetch provider resulted in a `mapstr.Put` on a `nil` `mapstr` and a race condition that can result in the fetch provider not be included in the context variable set.

## Why is it important?

<!-- Mandatory
Explain here the WHY, or the rationale/motivation for the changes.
-->

Using any context provider that is a fetch context provider will cause the Elastic Agent to crash.

## Checklist

<!-- Mandatory
Add a checklist of things that are required to be reviewed in order to have the PR approved

List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply. (Just in case, strikethrough uses two tildes. ~~Scratch this.~~)
-->

- [x] I have read and understood the [pull request guidelines](https://github.com/elastic/elastic-agent/blob/main/CONTRIBUTING.md#pull-request-guidelines) of this project.
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- ~~[ ] I have made corresponding changes to the documentation~~
- ~~[ ] I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- ~~[ ] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)~~
- ~~[ ] I have added an integration test or an E2E test~~

## Disruptive User Impact

<!--
Will the changes introduced by this PR cause disruption to users in any way? If so, please describe what changes users
could make on their end to nullify or minimize this disruption. Consider impacts in related systems, not just directly
when using Elastic Agent.
-->

None

## Related issues

<!-- Recommended
Link related issues below. Insert the issue link or reference after the word "Closes" if merging this should automatically close it.

- Closes #123
- Relates #123
- Requires #123
- Superseds #123
-->
- Closes #6957

elasticmachine commented:
Pinging @elastic/elastic-agent-control-plane (Team:Elastic-Agent-Control-Plane)

pkoutsovasilis reviewed (commented):
I did test this PR change and I can confirm that I didn't experience the reported panic with it. However I did observe that sometimes (actually most of them) the `Fetch` method of the provider is not called at all. Looked like a race condition but I can't pin it down to a specific place. With your knowledge around these code parts @blakerouse could you see a flow where [this code](https://github.com/elastic/elastic-agent/blob/800d19e780fc03383ab4978c8809ff96428fbb1c/internal/pkg/agent/transpiler/vars.go#L95) is missing a provider due to race conditions?

pkoutsovasilis commented:
the path so far I have found to make Fetch not getting called is by adding an artificial delay of 1 sec in the "constructor" of the kubernetes_secrets provider [here](https://github.com/elastic/elastic-agent/blob/main/internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets.go#L83). This makes me suspect that there is somewhere a `<-time.After(...)` with a time limit that when exceeded a provider is lost in the void. To last bit is just a suspicion at the moment I don't have code proof. Looking more into that

swiatekm commented:
It's definitely possible for the observe call [here](https://github.com/elastic/elastic-agent/blob/93b3138a5a4222054f23e9600ca8ed0a9995d865/internal/pkg/agent/application/coordinator/coordinator.go#L1353) to return without all the providers having been started. Providers are started asynchronously, and the fetch providers map is also updated asynchronously [here](https://github.com/elastic/elastic-agent/blob/93b3138a5a4222054f23e9600ca8ed0a9995d865/internal/pkg/composable/controller.go#L200), but this doesn't trigger new variables. So if this happens, and nothing triggers a new config reload in the coordinator, we'll never get variables with the new providers.

Does that sound possible @blakerouse ?

blakerouse commented:
@swiatekm It is not that the providers need to be fully started, they publish a new variable once fully started which will signal a new set of variables including the contents.

The second link is defiantly more related to the issue. The issue does seem to be that the fetch providers are not sent over the channel until they are initialized if that takes time (which kubernetes_secrets will to create the client to kubernetes) then it will not be included in the vars.

I have been working on a proper fix, just wanted to inform that I know the issue just not at the proper solution just yet.

blakerouse commented:
@pkoutsovasilis @swiatekm I have updated this to handle the case for the fetch providers not always being present. I actually like this much better as it removes a lock. I was thinking in my mind this is what it was already doing, but I was incorrect. This change works because each context provider is always started in its own goroutine allowing for a blocking channel to be used for sending the fetch providers. Now a fetch provider results in a new observed set of variables, which is what should have been occurring.

The delay added in the test (which I added a comment on), ensures this code is working correctly. Without the change and the delay present I was seeing what @pkoutsovasilis was describing.

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/elastic-agent/builds/17595)
* Commit: 01f309483ab61b28f7c08d2eae42e5edd6d8b48e

### Failed CI Steps
* [Unit tests - macOS 15 ARM](https://buildkite.com/elastic/elastic-agent/builds/17595#01954411-5a89-45ac-bd01-34a68551428e)

cc @blakerouse

<!--buildkite-pr-comment-elastic-agent
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-agent/builds/17595","number":17595,"commit":"01f309483ab61b28f7c08d2eae42e5edd6d8b48e"}],"number":17595}
buildkite-pr-comment-->

pkoutsovasilis reviewed (commented):
yes I just tried again reproducing that but with the latest commits I am not able to do so. Code changes LGTM but I will have to wait for our CI to become again healthy to make sure nothing somewhere is affected by the changes. That said, I wonder if the race condition issue about the providers should be backported also to version prior to 8.18, wdyt? 🤔

blakerouse commented:
buildkite test this

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/elastic-agent/builds/17639)
* Commit: 01f309483ab61b28f7c08d2eae42e5edd6d8b48e

### Failed CI Steps
* [Unit tests - macOS 15 ARM](https://buildkite.com/elastic/elastic-agent/builds/17639#019547d6-b345-4e11-8228-ab4a739fb5b5)

### History
* :broken_heart: [Build #17595](https://buildkite.com/elastic/elastic-agent/builds/17595) failed 01f309483ab61b28f7c08d2eae42e5edd6d8b48e

cc @blakerouse

<!--buildkite-pr-comment-elastic-agent
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-agent/builds/17639","number":17639,"commit":"01f309483ab61b28f7c08d2eae42e5edd6d8b48e"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-agent/builds/17595","number":17595,"commit":"01f309483ab61b28f7c08d2eae42e5edd6d8b48e"}],"number":17639}
buildkite-pr-comment-->

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/elastic-agent/builds/17654)
* Commit: 6b07f2281adb8bede55ecc1ca5ed98dfb4c237ca

### Failed CI Steps
* [Unit tests - macOS 15 ARM](https://buildkite.com/elastic/elastic-agent/builds/17654#019548e5-ecbd-4a98-8a77-9ea7e52e9104)

### History
* :broken_heart: [Build #17639](https://buildkite.com/elastic/elastic-agent/builds/17639) failed 01f309483ab61b28f7c08d2eae42e5edd6d8b48e
* :broken_heart: [Build #17595](https://buildkite.com/elastic/elastic-agent/builds/17595) failed 01f309483ab61b28f7c08d2eae42e5edd6d8b48e

cc @blakerouse

<!--buildkite-pr-comment-elastic-agent
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-agent/builds/17654","number":17654,"commit":"6b07f2281adb8bede55ecc1ca5ed98dfb4c237ca"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-agent/builds/17639","number":17639,"commit":"01f309483ab61b28f7c08d2eae42e5edd6d8b48e"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-agent/builds/17595","number":17595,"commit":"01f309483ab61b28f7c08d2eae42e5edd6d8b48e"}],"number":17654}
buildkite-pr-comment-->

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/elastic-agent/builds/17669)
* Commit: 59e42d1d815af75b28316e994bad79ae2d0ee11d

### Failed CI Steps
* [x86_64:sudo: fleet-airgapped](https://buildkite.com/elastic/elastic-agent/builds/17669#0195490e-d4f5-4f19-900e-b20d6a4c3f4a)

### History
* :broken_heart: [Build #17654](https://buildkite.com/elastic/elastic-agent/builds/17654) failed 6b07f2281adb8bede55ecc1ca5ed98dfb4c237ca
* :broken_heart: [Build #17639](https://buildkite.com/elastic/elastic-agent/builds/17639) failed 01f309483ab61b28f7c08d2eae42e5edd6d8b48e
* :broken_heart: [Build #17595](https://buildkite.com/elastic/elastic-agent/builds/17595) failed 01f309483ab61b28f7c08d2eae42e5edd6d8b48e

cc @blakerouse

<!--buildkite-pr-comment-elastic-agent
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-agent/builds/17669","number":17669,"commit":"59e42d1d815af75b28316e994bad79ae2d0ee11d"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-agent/builds/17654","number":17654,"commit":"6b07f2281adb8bede55ecc1ca5ed98dfb4c237ca"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-agent/builds/17639","number":17639,"commit":"01f309483ab61b28f7c08d2eae42e5edd6d8b48e"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-agent/builds/17595","number":17595,"commit":"01f309483ab61b28f7c08d2eae42e5edd6d8b48e"}],"number":17669}
buildkite-pr-comment-->
